add ca_path argument to uri module

heiderich opened the pull request:
##### SUMMARY

Until now the uri module does not support custom CA certificate files (see https://stackoverflow.com/questions/47890394/ansible-uri-module-with-cacert-option). The changes in this PR allow using a https endpoint that uses a self-signed certificate by specifying the CA certificate file.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
uri

##### ADDITIONAL INFORMATION

<!--- Paste verbatim command output below, e.g. before and after your change -->

Before the change:

```
$ cat test_uri_without_ca_path.yml 
---
- hosts: localhost
  tasks:
    - uri:
        url: "https://self-signed.badssl.com/"
```

```
$ ansible-playbook -D test_uri_without_ca_path.yml 
[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

PLAY [localhost] *********************************************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ***************************************************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [uri] ***************************************************************************************************************************************************************************************************************************************************
fatal: [localhost]: FAILED! => {"changed": false, "elapsed": 0, "msg": "Status code was -1 and not [200]: Request failed: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate (_ssl.c:1108)>", "redirected": false, "status": -1, "url": "https://self-signed.badssl.com/"}

PLAY RECAP ***************************************************************************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   
```

After the change:
```
$ cat test_uri_with_ca_path.yml
---
- hosts: localhost
  tasks:
    - uri:
        url: "https://self-signed.badssl.com/"
        ca_path: "badssl-com-chain.pem"
```

```
$ cat badssl-com-chain.pem
-----BEGIN CERTIFICATE-----
MIIDeTCCAmGgAwIBAgIJAPziuikCTox4MA0GCSqGSIb3DQEBCwUAMGIxCzAJBgNVBAYTAlVTMRMwEQYDVQQIDApDYWxpZm9ybmlhMRYwFAYDVQQHDA1TYW4gRnJhbmNpc2NvMQ8wDQYDVQQKDAZCYWRTU0wxFTATBgNVBAMMDCouYmFkc3NsLmNvbTAeFw0xOTEwMDkyMzQxNTJaFw0yMTEwMDgyMzQxNTJaMGIxCzAJBgNVBAYTAlVTMRMwEQYDVQQIDApDYWxpZm9ybmlhMRYwFAYDVQQHDA1TYW4gRnJhbmNpc2NvMQ8wDQYDVQQKDAZCYWRTU0wxFTATBgNVBAMMDCouYmFkc3NsLmNvbTCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBAMIE7PiM7gTCs9hQ1XBYzJMY61yoaEmwIrX5lZ6xKyx2PmzAS2BMTOqytMAPgLaw+XLJhgL5XEFdEyt/ccRLvOmULlA3pmccYYz2QULFRtMWhyefdOsKnRFSJiFzbIRMeVXk0WvoBj1IFVKtsyjbqv9u/2CVSndrOfEk0TG23U3AxPxTuW1CrbV8/q71FdIzSOciccfCFHpsKOo3St/qbLVytH5aohbcabFXRNsKEqveww9HdFxBIuGa+RuT5q0iBikusbpJHAwnnqP7i/dAcgCskgjZjFeEU4EFy+b+a1SYQCeFxxC7c3DvaRhBB0VVfPlkPz0sw6l865MaTIbRyoUCAwEAAaMyMDAwCQYDVR0TBAIwADAjBgNVHREEHDAaggwqLmJhZHNzbC5jb22CCmJhZHNzbC5jb20wDQYJKoZIhvcNAQELBQADggEBAGlwCdbPxflZfYOaukZGCaxYK6gpincX4Lla4Ui2WdeQxE95w7fChXvP3YkE3UYUE7mupZ0eg4ZILr/A0e7JQDsgIu/SRTUE0domCKgPZ8v99k3Avka4LpLK51jHJJK7EFgo3ca2nldd97GM0MU41xHFk8qaK1tWJkfrrfcGwDJ4GQPIiLlm6i0yHq1Qg1RypAXJy5dTlRXlCLd8ufWhhiwW0W75Va5AEnJuqpQrKwl3KQVewGj67WWRgLfSr+4QG1mNvCZb2CkjZWmxkGPuoP40/y7Yu5OFqxP5tAjj4YixCYTWEVA0pmzIzgBg+JIe3PdRy27T0asgQW/F4TY61Yk=
-----END CERTIFICATE-----
```

```
$ ansible-playbook -D test_uri_with_ca_path.yml 
[WARNING]: You are running the development version of Ansible. You should only run Ansible from "devel" if you are modifying the Ansible engine, or trying out features under development. This is a rapidly changing source of code and can become unstable
at any point.
[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

PLAY [localhost] *********************************************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ***************************************************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [uri] ***************************************************************************************************************************************************************************************************************************************************
[WARNING]: The value "status_code: 200" (type int) was converted to "status_code: '200'" (type string). If this does not look like what you expect, quote the entire value to ensure it does not change.
ok: [localhost]

PLAY RECAP ***************************************************************************************************************************************************************************************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   

```

I am not sure whether there is a chance to get this into ansible 2.10. If so, feel free to change the `version_added: '2.11'` line in `lib/ansible/modules/uri.py`.

heiderich committed:
add ca_path argument to uri module

samdoran reviewed (changes requested):
Please create a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment) and integration tests. See [this fragment](https://github.com/ansible/ansible/blob/c2f528b7686c9573aa8493357219e0bbe4416dd3/changelogs/fragments/reboot-change-default-boot-command.yaml) as an example.

heiderich committed:
add changelog fragment for #71979 (ca_path for uri)

bcoca commented:
the changelog should be added to THIS PR/branch

heiderich committed:
use unix line endings

heiderich commented:
Thank you for your comments @samdoran and @bcoca!. I added a changelog fragment in 4ea2b3c11743d8d66f621a2e77b5ee20d5ef841f.

Commit 72cc6fc3a0233e8b6a6456eab7880157db7989a6 contains integration tests. I checked in the CA chain used by https://self-signed.badssl.com/. This is not ideal, as it may change in the future. However I did not find openssl in the test container to retrieve it. Are the acceptance tests in the current form acceptable? If not, can we install openssl in the test containers?

I will have to investigate why one test fails (https://app.shippable.com/github/ansible/ansible/runs/174604/68/console).

sivel commented:
Our http-test-container has some badssl endpoints configured.

https://github.com/ansible/ansible/blob/3e9943bc5e7a9cd393757aa8100d7fed80bd316e/test/integration/targets/uri/tasks/main.yml#L91-L96

https://github.com/ansible/http-test-container/blob/master/nginx.sites.conf

If we need some extra endpoints, we will need to configure them in the http-test-container, instead of relying on external resources.

As of now the http-test-container has it's own CA, and we install it globally, so maybe we could just do that.  Set a path to an invalid CA, and then set an explicit path to the valid CA.  The `prepare_http_tests` target has the tasks related to setting up the ca bundles etc...

heiderich committed:
make uri ca_path tests use self_signed_host

heiderich committed:
delete test/integration/targets/uri/files/badssl-com-chain.pem

heiderich commented:
Thanks for your feedback @sivel. I was expecting someone to come up with the argument that we should better not depend on external services. I totally agree and was just too lazy to understand and adopt the internal http infrastructure.

The CA certificate from `http-test-container` is installed in the test container by `prepare_http_tests`. As the https endpoints of `http-test-container` are signed by that CA, they can all be validated without using the new `ca_path` argument. To properly test that, we instead need a https endpoint that uses a certificate that is not signed by any of the "known" CAs of the test container (which includes the CA certs of `http-test-container`). Therefore I did the following:

- Create a second CA in http-test-container under `/root/ca2` and create a new https endpoint with hostname `self-signed.ansible.http.tests` that is signed by that CA (see https://github.com/ansible/http-test-container/pull/7).
- The integration tests for ca_path now use the new `self-signed.ansible.http.tests` endpoint.
    - Without providing the ca_path argument the ssl validation fails (as I do not install the cert of the new ca2 in the test container.
    - I download the cert of the ca2 in a temporary directory and use it as the argument of `ca_path` to successfully verify the cert of `self-signed.ansible.http.tests`.

**Note:** The tests will only pass once my changes in `http-test-container` are merged and released as version 1.2.0 (I voluntarily choose this tag already in https://github.com/ansible/ansible/pull/71979/commits/308cf01f00f7e6adcef8c2bdcbf3b8567a8ae15b).

samdoran commented on changelogs/fragments/71979_ca_path_for_uri.yaml:
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - add ``ca_path`` argument to ``uri`` module to allow specification of a CA certificate (https://github.com/ansible/ansible/pull/71979).
```
```suggestion
  - uri - add ``ca_path`` argument to allow specification of a CA certificate (https://github.com/ansible/ansible/pull/71979).
```

samdoran commented on lib/ansible/modules/uri.py:
```diff
@@ -625,6 +630,7 @@ def main():
         headers=dict(type='dict', default={}),
         unix_socket=dict(type='path'),
         remote_src=dict(type='bool', default=False),
+        ca_path=dict(type='path', default=None)
```
```suggestion
        ca_path=dict(type='path', default=None),
```

samdoran commented:
I merged https://github.com/ansible/http-test-container/pull/7 and created a 1.2.0 of the image release.

heiderich committed:
Update changelogs/fragments/71979_ca_path_for_uri.yaml

Co-authored-by: Sam Doran <sdoran@redhat.com>

heiderich committed:
Update lib/ansible/modules/uri.py

Co-authored-by: Sam Doran <sdoran@redhat.com>

heiderich commented:
In order to understand the failed tests https://app.shippable.com/github/ansible/ansible/runs/174946/53/console
and https://app.shippable.com/github/ansible/ansible/runs/174946/68/console I executed

```ansible-test integration uri --docker centos6 -vvv```

The relevant part of the output is:

```
TASK [uri : test https fetch to a site with self signed certificate using ca_path] ***
task path: /root/ansible/test/results/.tmp/integration/uri-I98yNX-ÅÑŚÌβŁÈ/test/integration/targets/uri/tasks/main.yml:139
<testhost> ESTABLISH LOCAL CONNECTION FOR USER: root
<testhost> EXEC /bin/sh -c 'echo ~root && sleep 0'
<testhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp `"&& mkdir "` echo /root/.ansible/tmp/ansible-tmp-1602177823.27-3515-227590889238505 `" && echo ansible-tmp-1602177823.27-3515-227590889238505="` echo /root/.ansible/tmp/ansible-tmp-1602177823.27-3515-227590889238505 `" ) && sleep 0'
Using module file /root/ansible/lib/ansible/modules/uri.py
<testhost> PUT /root/.ansible/tmp/ansible-local-399UOPkw2/tmpSE_Tjr TO /root/.ansible/tmp/ansible-tmp-1602177823.27-3515-227590889238505/AnsiballZ_uri.py
<testhost> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1602177823.27-3515-227590889238505/ /root/.ansible/tmp/ansible-tmp-1602177823.27-3515-227590889238505/AnsiballZ_uri.py && sleep 0'
<testhost> EXEC /bin/sh -c '/tmp/python-q_mBFa-ansible/python /root/.ansible/tmp/ansible-tmp-1602177823.27-3515-227590889238505/AnsiballZ_uri.py && sleep 0'
<testhost> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-tmp-1602177823.27-3515-227590889238505/ > /dev/null 2>&1 && sleep 0'
The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_ansible.legacy.uri_payload_fzqKTg/ansible_ansible.legacy.uri_payload.zip/ansible/module_utils/urls.py", line 1619, in fetch_url
    unix_socket=unix_socket, ca_path=ca_path)
  File "/tmp/ansible_ansible.legacy.uri_payload_fzqKTg/ansible_ansible.legacy.uri_payload.zip/ansible/module_utils/urls.py", line 1401, in open_url
    unredirected_headers=unredirected_headers)
  File "/tmp/ansible_ansible.legacy.uri_payload_fzqKTg/ansible_ansible.legacy.uri_payload.zip/ansible/module_utils/urls.py", line 1306, in open
    return urllib_request.urlopen(request, None, timeout)
  File "/usr/lib64/python2.6/urllib2.py", line 126, in urlopen
    return _opener.open(url, data, timeout)
  File "/usr/lib64/python2.6/urllib2.py", line 389, in open
    req = meth(req)
  File "/tmp/ansible_ansible.legacy.uri_payload_fzqKTg/ansible_ansible.legacy.uri_payload.zip/ansible/module_utils/urls.py", line 1008, in http_request
    ssl_s = ssl.wrap_socket(s, ca_certs=tmp_ca_cert_path, cert_reqs=ssl.CERT_REQUIRED, ssl_version=PROTOCOL)
  File "/usr/lib64/python2.6/ssl.py", line 342, in wrap_socket
    suppress_ragged_eofs=suppress_ragged_eofs)
  File "/usr/lib64/python2.6/ssl.py", line 118, in __init__
    cert_reqs, ssl_version, ca_certs)
TypeError: sslwrap() argument 7 must be string or None, not list
fatal: [testhost]: FAILED! => {
    "changed": false, 
    "elapsed": 0, 
    "invocation": {
        "module_args": {
            "attributes": null, 
            "body": null, 
            "body_format": "raw", 
            "ca_path": "/tmp/ansible.eL8MYU.test/ca2cert.pem", 
            "client_cert": null, 
            "client_key": null, 
            "creates": null, 
            "dest": "/root/ansible/test/results/.tmp/output_dir/self-signed_using_ca_path.html", 
            "follow_redirects": "safe", 
            "force": false, 
            "force_basic_auth": false, 
            "group": null, 
            "headers": {}, 
            "http_agent": "ansible-httpget", 
            "method": "GET", 
            "mode": null, 
            "owner": null, 
            "remote_src": false, 
            "removes": null, 
            "return_content": false, 
            "selevel": null, 
            "serole": null, 
            "setype": null, 
            "seuser": null, 
            "src": null, 
            "status_code": [
                200
            ], 
            "timeout": 30, 
            "unix_socket": null, 
            "unsafe_writes": false, 
            "url": "https://self-signed.ansible.http.tests/", 
            "url_password": null, 
            "url_username": null, 
            "use_proxy": true, 
            "validate_certs": true
        }
    }, 
    "msg": "Status code was -1 and not [200]: An unknown error occurred: sslwrap() argument 7 must be string or None, not list", 
    "path": "/root/ansible/test/results/.tmp/output_dir/self-signed_using_ca_path.html", 
    "redirected": false, 
    "status": -1, 
    "url": "https://self-signed.ansible.http.tests/"
}

```

Observations:

1) The function `SSLValidationHandler.get_ca_cert()` (https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/urls.py#L810) returns a list of length 1 that contains a string with the content(s) of (a) certificate(s).
2) In the other case the first returned argument is a filepath as a string (https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/urls.py#L906).

It seems that at least  `_ssl.sslwrap` in Python 2.6 in centos6 has a problem with 1.

@sivel, you implemented https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/urls.py#L818-L829. 

Can you suggest what to do here?

sivel commented:
I haven't looked it in depth, but it looks like the return on 829 should be:

```
return self.ca_path, cadata, paths_checked
```

827 and 828 can probably be removed as well in this case.

heiderich committed:
return path of ca cert instead of its content

heiderich committed:
connect to port 444 on self_signed_host

and use quay.io/ansible/http-test-container:1.3.0

heiderich commented:
Commit 5220ff3510a027d7504d7e9d5f1c62d7077302eb contains the proposed changes. Now running the integration tests with

```
ansible-test integration uri --docker centos6 -vvv
```

fails with the following error:

```
TASK [uri : test https fetch to a site with self signed certificate using ca_path] ***
task path: /root/ansible/test/results/.tmp/integration/uri-b4k6rn-ÅÑŚÌβŁÈ/test/integration/targets/uri/tasks/main.yml:139
<testhost> ESTABLISH LOCAL CONNECTION FOR USER: root
<testhost> EXEC /bin/sh -c 'echo ~root && sleep 0'
<testhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp `"&& mkdir "` echo /root/.ansible/tmp/ansible-tmp-1602264286.45-3517-262932260843115 `" && echo ansible-tmp-1602264286.45-3517-262932260843115="` echo /root/.ansible/tmp/ansible-tmp-1602264286.45-3517-262932260843115 `" ) && sleep 0'
Using module file /root/ansible/lib/ansible/modules/uri.py
<testhost> PUT /root/.ansible/tmp/ansible-local-400xD9zm7/tmpCp6crf TO /root/.ansible/tmp/ansible-tmp-1602264286.45-3517-262932260843115/AnsiballZ_uri.py
<testhost> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1602264286.45-3517-262932260843115/ /root/.ansible/tmp/ansible-tmp-1602264286.45-3517-262932260843115/AnsiballZ_uri.py && sleep 0'
<testhost> EXEC /bin/sh -c '/tmp/python-HEKq9n-ansible/python /root/.ansible/tmp/ansible-tmp-1602264286.45-3517-262932260843115/AnsiballZ_uri.py && sleep 0'
<testhost> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-tmp-1602264286.45-3517-262932260843115/ > /dev/null 2>&1 && sleep 0'
The full traceback is:
WARNING: The below traceback may *not* be related to the actual failure.
  File "/tmp/ansible_ansible.legacy.uri_payload_ikelaJ/ansible_ansible.legacy.uri_payload.zip/ansible/module_utils/urls.py", line 1617, in fetch_url
    unix_socket=unix_socket, ca_path=ca_path)
  File "/tmp/ansible_ansible.legacy.uri_payload_ikelaJ/ansible_ansible.legacy.uri_payload.zip/ansible/module_utils/urls.py", line 1399, in open_url
    unredirected_headers=unredirected_headers)
  File "/tmp/ansible_ansible.legacy.uri_payload_ikelaJ/ansible_ansible.legacy.uri_payload.zip/ansible/module_utils/urls.py", line 1304, in open
    return urllib_request.urlopen(request, None, timeout)
  File "/usr/lib64/python2.6/urllib2.py", line 126, in urlopen
    return _opener.open(url, data, timeout)
  File "/usr/lib64/python2.6/urllib2.py", line 389, in open
    req = meth(req)
  File "/tmp/ansible_ansible.legacy.uri_payload_ikelaJ/ansible_ansible.legacy.uri_payload.zip/ansible/module_utils/urls.py", line 1012, in http_request
    build_ssl_validation_error(self.hostname, self.port, paths_checked, e)
  File "/tmp/ansible_ansible.legacy.uri_payload_ikelaJ/ansible_ansible.legacy.uri_payload.zip/ansible/module_utils/urls.py", line 783, in build_ssl_validation_error
    raise SSLValidationError(' '.join(msg) % (hostname, port, ", ".join(paths)))
fatal: [testhost]: FAILED! => {
    "changed": false, 
    "invocation": {
        "module_args": {
            "attributes": null, 
            "body": null, 
            "body_format": "raw", 
            "ca_path": "/tmp/ansible.pAUAV6.test/ca2cert.pem", 
            "client_cert": null, 
            "client_key": null, 
            "creates": null, 
            "dest": "/root/ansible/test/results/.tmp/output_dir/self-signed_using_ca_path.html", 
            "follow_redirects": "safe", 
            "force": false, 
            "force_basic_auth": false, 
            "group": null, 
            "headers": {}, 
            "http_agent": "ansible-httpget", 
            "method": "GET", 
            "mode": null, 
            "owner": null, 
            "remote_src": false, 
            "removes": null, 
            "return_content": false, 
            "selevel": null, 
            "serole": null, 
            "setype": null, 
            "seuser": null, 
            "src": null, 
            "status_code": [
                200
            ], 
            "timeout": 30, 
            "unix_socket": null, 
            "unsafe_writes": false, 
            "url": "https://self-signed.ansible.http.tests/", 
            "url_password": null, 
            "url_username": null, 
            "use_proxy": true, 
            "validate_certs": true
        }
    }, 
    "msg": "Failed to validate the SSL certificate for self-signed.ansible.http.tests:443. Make sure your managed systems have a valid CA certificate installed. If the website serving the url uses SNI you need python >= 2.7.9 on your managed machine  (the python executable used (/tmp/python-HEKq9n-ansible/python) is version: 2.6.6 (r266:84292, Jun 20 2019, 14:14:55) [GCC 4.4.7 20120313 (Red Hat 4.4.7-23)]) or you can install the `urllib3`, `pyOpenSSL`, `ndg-httpsclient`, and `pyasn1` python modules to perform SNI verification in python >= 2.6. You can use validate_certs=False if you do not need to confirm the servers identity but this is unsafe and not recommended. Paths checked for this platform: /tmp/ansible.pAUAV6.test/ca2cert.pem. The exception msg was: [Errno 1] _ssl.c:492: error:14090086:SSL routines:SSL3_GET_SERVER_CERTIFICATE:certificate verify failed.", 
    "status": -1, 
    "url": "https://self-signed.ansible.http.tests/"
}
```

So apparently the fact that SNI is used in the test container causes problems with python 2.6.6. Therefore I changed the SSL port of `self-signed.ansible.http.tests` in https://github.com/ansible/http-test-container/pull/9 from 443 to 444. With this change these tests pass in my local setup. I suggest to merge that PR, release the new image using a new tag. Then we can change the port in this PR and update the version of the http-test-container. I hope that this will make this test pass.

samdoran commented:
I merged https://github.com/ansible/http-test-container/pull/8 and published a 1.2.1 release.

jborean93 commented:
Just an FYI, the change in the port would require some changes in `ansible-test` to make sure the correct ports are now forwarded.

jborean93 commented on lib/ansible/modules/uri.py:
```diff
@@ -141,6 +141,11 @@
     version_added: '2.4'
+  ca_path:
+    description:
+      - PEM formatted file that contains a CA certificate to be used
```
It would be good if you could expand a bit on this, it sounds like it's used for HTTPS cert validation so stating that

* This is used for validation
* It can contain multiple certs (at least I think it should)

Also, why stop at `uri`, this should be added to `get_url` and the `url` lookup plugin to ensure they all offer the same features.

heiderich replied:
I slightly expanded the description in 339fef01ccc8367370b8c5145c2d4ea2cae82403.

I did not test whether `ca_path` can contain multiple certificates. I would only state this in the documentation if we test it. In order to do so, we would need yet another endpoint in the http-test-container. Do you have a use case for this?

Yes, we may add this to other modules as well. For the moment I focus on the uri module.

jborean93 replied:
> I did not test whether ca_path can contain multiple certificates. I would only state this in the documentation if we test it. In order to do so

That's fair enough, I was unsure if you had tested it before or not.

heiderich committed:
state that the certificate in ca_path is used for validation

jborean93 commented:
I apologise but I've just recently merged https://github.com/ansible/ansible/pull/72113 which is the cause of the conflicts. You don't need to change the httptester version as 1.3.0 include your changes that were present in 1.2.1.

samdoran commented:
!waffling needs_revision

eqrx commented:
Hey everybody. What is the status of PR, can I help somehow? :)

samdoran commented:
@eqrx It looks like the holdup is resolving test issues.

samdoran commented:
/azp run

mkrizek commented:
!waffling core_review

heiderich commented:
@eqrx, as @samdoran writes, it remains to fix some test issues. I made an attempt with https://github.com/ansible/http-test-container/pull/7 and https://github.com/ansible/http-test-container/pull/9 and my last commits here, but some tests still fail. Any help to resolve the remaining issues would be very welcome.

samdoran commented:
The test failure seems like something is not right with the PowerShell HTTP watcher.

samdoran commented:
It looks like only the Windows integration tests are failing:

```
ERROR: Command "ssh -q -i /root/ansible/test/results/.tmp/id_rsa -o BatchMode=yes -o ServerAliveCountMax=4 -o ServerAliveInterval=15 -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -fT -R 8080:ansible.http.tests:80 -R 8443:ansible.http.tests:443 8444:ansible.http.tests:444 -p 22 administrator@ec2-3-87-125-76.compute-1.amazonaws.com 'powershell.exe -NoProfile -ExecutionPolicy Bypass -File .\ansible-test-http-watcher-1611585755.26.ps1 -Hosts "ansible.http.tests|sni1.ansible.http.tests|fail.ansible.http.tests|self-signed.ansible.http.tests" -Verbose'" returned exit status 255.
```

@jborean93 Do you have any insight into the failure? Is this an issue with `windows-httptester.ps1` or maybe a conflict with port forwards?

jborean93 commented:
Turns out the new forwarded port option added https://github.com/ansible/ansible/pull/71979/files#diff-c138acaaea6f25d5343d82539ade5353d9111d162a9575962068388e48d7a944R769 was missing the preceding `-R`  argument causing ssh to just fail entirely. Have pushed a commit that fixes that up.

jborean93 commented:
@samdoran CI is green, just pending your review.

samdoran commented:
@jborean93 You win the Internet! Thanks for fixing that. Sorry I didn't see it myself.

Now that we have the tests passing, I actually forgot that the `ansible-test` changes need to be in a separate PR. That will need a changelog outlining the changes to `ansible-test`. Once that PR is merged, tests for this PR will pass. @heiderich Does that make sense? Would you be able to create that PR?

heiderich commented:
> Now that we have the tests passing, I actually forgot that the ansible-test changes need to be in a separate PR. That will need a changelog outlining the changes to ansible-test. Once that PR is merged, tests for this PR will pass. @heiderich Does that make sense? Would you be able to create that PR?

@samdoran Yes, this sounds reasonable. Just to be sure: You are referring to the following changes, right?
https://github.com/ansible/ansible/pull/71979/files#diff-30aae2f5f35b5d72fc3d82d3a719634795587c92efec25ef43157446425c5b79
https://github.com/ansible/ansible/pull/71979/files#diff-c138acaaea6f25d5343d82539ade5353d9111d162a9575962068388e48d7a944
https://github.com/ansible/ansible/pull/71979/files#diff-7f8f64fd6e8fea4dd83c7e06db1c2a6fbe7df074a2a3c6de3270d2bb8be1c881
https://github.com/ansible/ansible/pull/71979/files#diff-59857da083825557fc0fe2ea69d1f8309729a6743f6283310f85c7824826d1de)

I cherry-picked the commits that modify these files 4 in https://github.com/heiderich/ansible/commits/self_signed_ansible_http_test.

If this is what you mean, I will create a PR from that branch to https://github.com/ansible/ansible/commits/devel and add a changelog.

samdoran commented:
@heiderich Yes, that is correct. Go ahead and open a PR with that branch. Once that is merged, those commits can be removed from this PR and it can be merged.

heiderich committed:
add integration tests for ca_path in the uri module

heiderich commented:
@samdoran @jborean93 I removed the ansible-test related commits 1f3642a460 5b4eccde3e, 5bcd85c65c and what was https://github.com/ansible/ansible/pull/73463/commits/54e5ee8b43a6bf1a0992c839d0f4b3648bd69254 on this branch here before. They are merged into devel as a part of PR #73463.  I rebased this brach on the current devel branch (7fb7d138fb1f5e3977d800cd2a56c60a101713ee). Could you please review this again?

jborean93 reviewed (approved):
Looks good!

samdoran commented:
Thanks for all your work on this.
